daemon/devices_linux.go

```diff
@@ -0,0 +1,41 @@
+package daemon // import "github.com/docker/docker/daemon"
+
+import (
+	"fmt"
+
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/pkg/capabilities"
+	specs "github.com/opencontainers/runtime-spec/specs-go"
+)
+
+var deviceDrivers = map[string]*deviceDriver{}
+
+type deviceDriver struct {
+	capset     capabilities.Set
+	updateSpec func(*specs.Spec, *deviceInstance) error
+}
+
+type deviceInstance struct {
+	req          container.DeviceRequest
+	selectedCaps []string
+}
+
+func registerDeviceDriver(name string, d *deviceDriver) {
+	deviceDrivers[name] = d
+}
+
+func (daemon *Daemon) handleDevice(req container.DeviceRequest, spec *specs.Spec) error {
+	if req.Driver == "" {
+		for _, dd := range deviceDrivers {
+			if selected := dd.capset.Match(req.Capabilities); selected != nil {
```

Comment:
I could make it an OR list of ANDs as well instead of a map.

Reply:
Perhaps we should if this is a concern, so in that case the driver would report itself as;

```json
{
  "capabilities": [
    ["capA", capB"],
    ["capB", capC"]
  ]
}
```


Could even decide to make it just return a list of capabilities for each GPU (then we can even determine the number of GPUs available);

```json
{
  "capabilities": [
    ["capA", capB"],
    ["capA", capB"],
    ["capA", capB"],
    ["capA", capB"],
    ["capB", capC"]
  ]
}
```

But perhaps that breaks the abstraction

Reply:
I suggest we punt on the problem since the problem is extremely unlikely to happen at this time and the structure is that of the device driver so it's internal, we can change it. The API needs to be locked down.